packages/realm/src/schema/normalize.ts

```diff
@@ -43,163 +60,350 @@ export const PRIMITIVE_TYPES = new Set<PropertyTypeName>([
+  }
+
+  if (optionalIsImplicitlyTrue(type, objectType)) {
+    const displayed =
+      type === "mixed" || objectType === "mixed"
+        ? "'mixed' types"
+        : "User-defined types as standalone objects and in dictionaries";
+    assert(optional !== false, errMessage(info, `${displayed} are always optional and cannot be made non-optional.`));
+    optional = true;
+  } else if (optionalIsImplicitlyFalse(type, objectType)) {
+    assert(
+      optional !== true,
+      errMessage(info, "User-defined types in lists and sets are always non-optional and cannot be made optional."),
+    );
+    optional = false;
+  }
+
+  const normalizedSchema: CanonicalObjectSchemaProperty = {
+    name: info.propertyName,
+    type: type as PropertyTypeName,
+    optional: !!optional,
+    indexed: !!propertySchema.indexed,
+    mapTo: propertySchema.mapTo || info.propertyName,
+  };
+  // Add optional properties only if defined (tests expect no 'undefined' properties)
+  if (objectType !== undefined) normalizedSchema.objectType = objectType;
+  if (property !== undefined) normalizedSchema.property = property;
+  if (defaultValue !== undefined) normalizedSchema.default = defaultValue;
+
+  return normalizedSchema;
+}
+
+/**
+ * Determine whether a property always is implicitly optional (nullable).
+ */
+function optionalIsImplicitlyTrue(type: string, objectType: string | undefined): boolean {
+  return (
+    type === "mixed" ||
+    objectType === "mixed" ||
+    type === "object" ||
+    (type === "dictionary" && isUserDefined(objectType))
+  );
+}
+
+/**
+ * Determine whether a property always is implicitly non-optional (non-nullable).
+ */
+function optionalIsImplicitlyFalse(type: string, objectType: string | undefined): boolean {
+  return (type === "list" || type === "set" || type === "linkingObjects") && isUserDefined(objectType);
+}
+
+/**
+ * Determine whether a string ends with a shorthand collection ('[]' or '{}' or '<>').
+ */
+function hasCollectionSuffix(input: string): boolean {
+  const end = input.substring(input.length - 2);
+  return !!COLLECTION_SHORTHAND_TO_NAME[end];
+}
+
+/**
+ * Assert that shorthand notation is not being used.
+ */
+function assertNotUsingShorthand(input: string | undefined, info: PropertyInfo): void {
+  const shorthands: string[] = [];
+
+  if (input && hasCollectionSuffix(input)) {
+    shorthands.push(input.substring(input.length - 2));
+    input = input.substring(0, input.length - 2);
+  }
+
+  if (input?.endsWith("?")) {
+    shorthands.push("?");
   }
+
+  assert(
+    !shorthands.length,
+    errMessage(
+      info,
+      `Cannot use shorthand '${shorthands.join("' and '")}' in 'type' or 'objectType' when defining property objects.`,
+    ),
+  );
 }
 
+/**
+ * Get an error message for an invalid property schema.
+ */
+function errMessage({ objectName, propertyName }: PropertyInfo, message: string): string {
+  return `Invalid type declaration for property '${objectName}.${propertyName}': ${message}`;
```

Comment:
I'd consider not using the `.` here, to avoid confusion, since the property is never a static member of the object schema class. How would you feel about:

```ts
return `Invalid type declaration for property '${propertyName}' on '${objectName}': ${message}`;
```
or alternatively
```ts
return `Invalid type declaration for ${objectName}'s '${propertyName}' property: ${message}`;
```

Reply:
We can definitely change this, I think the 1st suggestion is most readable :)